kratos/utilities/strategies_utilities.h

```diff
@@ -0,0 +1,228 @@
+            case SCALING_DIAGONAL::NO_SCALING:
+                return 1.0;
+            case SCALING_DIAGONAL::CONSIDER_PRESCRIBED_DIAGONAL: {
+                const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+                KRATOS_ERROR_IF_NOT(r_current_process_info.Has(BUILD_SCALE_FACTOR)) << "Scale factor not defined at process info" << std::endl;
+                return r_current_process_info.GetValue(BUILD_SCALE_FACTOR);
+            }
+            case SCALING_DIAGONAL::CONSIDER_NORM_DIAGONAL:
+                return GetDiagonalNorm<TSparseSpace>(rA)/static_cast<double>(rA.size1());
+            case SCALING_DIAGONAL::CONSIDER_MAX_DIAGONAL:
+                return GetMaxDiagonal<TSparseSpace>(rA);
+//                 return TSparseSpace::TwoNorm(rA)/static_cast<double>(rA.size1());
+            default:
+                return GetMaxDiagonal<TSparseSpace>(rA);
+        }
+    }
+
+    /**
+     * @brief This method returns the diagonal norm considering for scaling the diagonal
+     * @param rA The LHS matrix
+     * @return The diagonal norm
+     */
+    template<class TSparseSpace>
+    static double GetDiagonalNorm(typename TSparseSpace::MatrixType& rA)
+    {
+        double* Avalues = rA.value_data().begin();
+        std::size_t* Arow_indices = rA.index1_data().begin();
+        std::size_t* Acol_indices = rA.index2_data().begin();
+
+        const double diagonal_norm = IndexPartition<std::size_t>(TSparseSpace::Size1(rA)).for_each<SumReduction<double>>([&](std::size_t Index){
+            const std::size_t col_begin = Arow_indices[Index];
+            const std::size_t col_end = Arow_indices[Index+1];
+            for (std::size_t j = col_begin; j < col_end; ++j) {
+                if (Acol_indices[j] == Index ) {
+                    return std::pow(Avalues[j], 2);
+                }
+            }
+            return 0.0;
+        });
+
+        return std::sqrt(diagonal_norm);
+    }
+
+    /**
+     * @brief This method returns the diagonal max value
+     * @param rA The LHS matrix
+     * @return The diagonal  max value
+     */
+    template<class TSparseSpace>
+    static double GetAveragevalueDiagonal(typename TSparseSpace::MatrixType& rA)
+    {
+        return 0.5 * (GetMaxDiagonal<TSparseSpace>(rA) + GetMinDiagonal<TSparseSpace>(rA));
+    }
+
+    /**
+     * @brief This method returns the diagonal max value
+     * @param rA The LHS matrix
+     * @return The diagonal  max value
+     */
+    template<class TSparseSpace>
+    static double GetMaxDiagonal(typename TSparseSpace::MatrixType& rA)
+    {
+        double* Avalues = rA.value_data().begin();
+        std::size_t* Arow_indices = rA.index1_data().begin();
+        std::size_t* Acol_indices = rA.index2_data().begin();
+
+        return IndexPartition<std::size_t>(TSparseSpace::Size1(rA)).for_each<MaxReduction<double>>([&](std::size_t Index){
+            const std::size_t col_begin = Arow_indices[Index];
```

Comment:
just wondering (i don't know the answer) if for this it would not be faster to do the following

~~~cpp
return IndexPartition<std::size_t>(TSparseSpace::Size1(rA)).for_each<MaxReduction<double>>([&](std::size_t Index){
            const std::size_t col_size = Arow_indices[Index+1] - Arow_indices[Index];
            if(col_size>0) {
                    return std::abs(Avalues(i,i));
                }
            else
                   return std::numeric_limits<double>::max();
            }
        });
~~~

Reply:
Huuum, I can try

Reply:
? the (i,j) methos is there ...

Reply:
Maybe is because of the max, should be 0?

Reply:
Let me see

Reply:
Now it works

Reply:
the point is which one is faster (this is what i don't know)

Reply:
I think depends on the implementation of `A(i,j)`, worst case scenario should be the same, as we were just iterating one by one

Reply:
Nope, https://www.boost.org/doc/libs/1_67_0/boost/numeric/ublas/matrix_sparse.hpp, previous should be faster